hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -398,49 +370,132 @@ private void updateExpiryFields(
+            requireAdmin(context, token);
+            context.requireKeyOrThrow(op.treasuryOrThrow(), INVALID_ACCOUNT_ID);
         }
-        if (op.hasWipeKey()) {
-            validateTrue(originalToken.hasWipeKey(), TOKEN_HAS_NO_WIPE_KEY);
-            builder.wipeKey(op.wipeKey());
+        // updating auto-renew account needs admin key and the new auto-renewal account key
+        if (op.hasAutoRenewAccount()) {
+            requireAdmin(context, token);
+            context.requireKeyOrThrow(op.autoRenewAccountOrThrow(), INVALID_AUTORENEW_ACCOUNT);
         }
-        if (op.hasSupplyKey()) {
-            validateTrue(originalToken.hasSupplyKey(), TOKEN_HAS_NO_SUPPLY_KEY);
-            builder.supplyKey(op.supplyKey());
+        // updating admin key needs the old admin key and the new admin key
+        if (op.hasAdminKey()) {
+            requireAdmin(context, token);
+            context.requireKey(op.adminKeyOrThrow());
         }
-        if (op.hasFeeScheduleKey()) {
-            validateTrue(originalToken.hasFeeScheduleKey(), TOKEN_HAS_NO_FEE_SCHEDULE_KEY);
-            builder.feeScheduleKey(op.feeScheduleKey());
+        // Any key removal requires admin key
+        if (containsKeyRemoval(op)) {
+            requireAdmin(context, token);
         }
-        if (op.hasPauseKey()) {
-            validateTrue(originalToken.hasPauseKey(), TOKEN_HAS_NO_PAUSE_KEY);
-            builder.pauseKey(op.pauseKey());
+        // updating memo, name, symbol, auto-renew period needs admin key
+        if (updatesAdminOnlyNonKeyTokenProperty(op)) {
+            requireAdmin(context, token);
         }
-        if (op.hasMetadataKey()) {
-            validateTrue(originalToken.hasMetadataKey(), TOKEN_HAS_NO_METADATA_KEY);
-            builder.metadataKey(op.metadataKey());
+        // Any role key can be updated if the key is not empty of the token.
+        // Updating any non-admin keys with the key verification mode is NO_VALIDATION needs either
+        // admin key or the role key to sign.If the key verification mode is FULL_VALIDATION, then
+        // the new key also should sign the transaction.
+        for (final var tokenKey : NON_ADMIN_TOKEN_KEYS) {
+            if (tokenKey.isPresentInUpdate(op)) {
+                final var newRoleKey = tokenKey.getFromUpdate(op);
+                if (!isKeyRemoval(newRoleKey)) {
+                    if (op.keyVerificationMode() == NO_VALIDATION) {
+                        requireAdminOrRole(context, token, tokenKey);
+                    } else {
+                        // With "full" verification mode, our required key
+                        // structure is a 1/2 threshold with components:
+                        //   - Admin key
+                        //   - A 2/2 list including the role key and its replacement key
+                        final var key = tokenKey.getFromUpdate(op);
+                        requireAdminOrRole(context, token, tokenKey, key);
```

Comment:
WE are using 2/2 list just to follow the pattern right ? Its basically like needs (admin or role key ) and replacement key right ?

Reply:
We require _either_,
1. The `adminKey` _or_,
2. **Both** the `existingRoleKey` and `replacementRoleKey`

Reply:
The idea is that whenever the `adminKey` signs, no other key is required---because the `adminKey` has control to change those "lower keys" to whatever it wants.

Reply:
But how do we know the `replacementRoleKey` is valid ?

Reply:
Done in `pureChecks()`.

Reply:
We validate in `pureChecks`. Nvmd